libsql-replication/src/injector/hook.rs

```diff
@@ -0,0 +1,162 @@
+                self.is_txn = false;
+            }
+            tracing::trace!("applied frame batch");
+
+            Ok(())
+        } else {
+            tracing::error!("fatal replication error: failed to apply pages");
+            Err(())
+        }
+    }
+}
+
+/// Turn a list of `WalFrame` into a list of PgHdr.
+/// The caller has the responsibility to free the returned headers.
+/// return (headers, last_frame_no, size_after)
+fn make_page_header<'a>(frames: impl Iterator<Item = &'a FrameBorrowed>) -> (Headers<'a>, u32) {
+    let mut first_pg: *mut PgHdr = std::ptr::null_mut();
+    let mut current_pg;
+    let mut size_after = 0;
+
+    let mut headers_count = 0;
+    let mut prev_pg: *mut PgHdr = std::ptr::null_mut();
+    let mut frames = frames.peekable();
+    while let Some(frame) = frames.next() {
+        // the last frame in a batch marks the end of the txn
+        if frames.peek().is_none() {
+            size_after = frame.header().size_after;
+        }
+
+        let page = PgHdr {
+            pPage: std::ptr::null_mut(),
+            pData: frame.page().as_ptr() as _,
+            pExtra: std::ptr::null_mut(),
+            pCache: std::ptr::null_mut(),
+            pDirty: std::ptr::null_mut(),
+            pPager: std::ptr::null_mut(),
+            pgno: frame.header().page_no,
+            pageHash: 0,
+            flags: 0x02, // PGHDR_DIRTY - it works without the flag, but why risk it
+            nRef: 0,
+            pDirtyNext: std::ptr::null_mut(),
+            pDirtyPrev: std::ptr::null_mut(),
+        };
+        headers_count += 1;
+        current_pg = Box::into_raw(Box::new(page));
+        if first_pg.is_null() {
+            first_pg = current_pg;
+        }
+        if !prev_pg.is_null() {
+            unsafe {
+                (*prev_pg).pDirty = current_pg;
+            }
+        }
+        prev_pg = current_pg;
+    }
+
+    tracing::trace!("built {headers_count} page headers");
+
+    let headers = unsafe { Headers::new(first_pg) };
+    (headers, size_after)
+}
+
+init_static_wal_method!(INJECTOR_METHODS, InjectorHook);
+
+/// The injector hook hijacks a call to xframes, and replace the content of the call with it's own
+/// frames.
+/// The Caller must first call `set_frames`, passing the frames to be injected, then trigger a call
+/// to xFrames from the libsql connection (see dummy write in `injector`), and can then collect the
+/// result on the injection with `take_result`
+pub enum InjectorHook {}
+
+unsafe impl WalHook for InjectorHook {
+    type Context = InjectorHookCtx;
+
+    fn on_frames(
+        wal: &mut Wal,
+        _page_size: c_int,
+        _page_headers: *mut PgHdr,
+        _size_after: u32,
+        _is_commit: c_int,
+        sync_flags: c_int,
+        orig: XWalFrameFn,
+    ) -> c_int {
+        let wal_ptr = wal as *mut _;
+        let ctx = Self::wal_extract_ctx(wal);
+        let ret = ctx.inject_pages(sync_flags, orig, wal_ptr);
+        if ret.is_err() {
```

Comment:
let's tracing::error here to preserve the error context?

Reply:
this error is a `()`, and is directly propagated, it's caught later by the replicator. The actual error is already logged in `inject_pages`